spdx-utils/src/test/kotlin/SpdxExpressionTest.kt

```diff
@@ -367,6 +368,118 @@ class SpdxExpressionTest : WordSpec() {
+                val choice = "b".toSpdx()
+
+                shouldThrow<InvalidLicenseChoiceException> { expression.applyChoice(choice) }
+            }
+
+            "return the choice if the subExpression contains multiple choices" {
+                val expression = "a OR b OR c".toSpdx()
+                val choice = "a".toSpdx()
+
+                val result = expression.applyChoice(choice)
+
+                result shouldBe "a".toSpdx()
+            }
+
+            "return the new expression if only a part of the expression is matched by the subExpression" {
+                val expression = "a OR b OR c".toSpdx()
+                val choice = "b".toSpdx()
+                val subExpression = "a OR b".toSpdx()
+
+                val result = expression.applyChoice(choice, subExpression)
+
+                result shouldBe "b OR c".toSpdx()
+            }
+
+            "work with choices that itself are a choice" {
+                val expression = "a OR b OR c OR d".toSpdx()
+                val choice = "a OR b".toSpdx()
+                val subExpression = "a OR b OR c".toSpdx()
+
+                val result = expression.applyChoice(choice, subExpression)
+
+                result shouldBe "a OR b OR d".toSpdx()
+            }
+
+            "apply the choice if the expression contains multiple choices" {
+                val expression = "a OR b OR c".toSpdx()
+                val choice = "b".toSpdx()
+
+                val result = expression.applyChoice(choice)
+
+                result shouldBe "b".toSpdx()
+            }
+
+            "throw an exception if the chosen license is not an valid option" {
+                val expression = "a OR b".toSpdx()
+                val choice = "c".toSpdx()
+
+                shouldThrow<InvalidLicenseChoiceException> { expression.applyChoice(choice) }
+            }
+
+            "apply the choice if the expression is not in DNF" {
+                val expression = "(a OR b) AND c".toSpdx()
+                val choice = "a AND c".toSpdx()
+
+                val result = expression.applyChoice(choice)
+
+                result shouldBe "a AND c".toSpdx()
+            }
+
+            "return the reduced subExpression in DNF if the choice was valid" {
+                val expression = "(a OR b) AND c AND (d OR e)".toSpdx()
+                val choice = "a AND c AND d".toSpdx()
+                val subExpression = "a AND c AND d OR a AND c AND e".toSpdx()
+
+                val result = expression.applyChoice(choice, subExpression)
+
+                result shouldBe "a AND c AND d OR b AND c AND d OR b AND c AND e".toSpdx()
+            }
+
+            "return the expression if the subExpression does not match" {
+                val expression = "a OR b OR c".toSpdx()
+                val choice = "x".toSpdx()
+                val subExpression = "x OR y OR z".toSpdx()
+
+                val result = expression.applyChoice(choice, subExpression)
```

Comment:
Shouldn't it throw an exception if the subexpression is invalid like for invalid license choices?

Reply:
Yes I agree that this would make the API more consistent and will fail early if the developer uses this API in another way as it was intended.